File: day1/part2/solution.rb

def solve(rotations)
  current_position = 50
  total_zero_count = 0

  rotations.each do |line|
    next if line.empty?

    direction = line[0]
    distance = line[1..-1].to_i

    zeros_this_turn = 0
    if direction == 'R'
      # 'div' in Ruby is floor division
      zeros_this_turn = (current_position + distance).div(100) - current_position.div(100)
    elsif direction == 'L'
      zeros_this_turn = (current_position - 1).div(100) - (current_position - distance - 1).div(100)
    end
    
    total_zero_count += zeros_this_turn

    if direction == 'R'
      current_position += distance
    elsif direction == 'L'
      current_position -= distance
    end
  end
  total_zero_count
end

def main
  input_file_path = File.join(__dir__, '../part1/input1.txt')
  rotations = File.readlines(input_file_path, chomp: true) # chomp: true removes newlines

  password = solve(rotations)
  puts "The password is: #{password}"
end

main if __FILE__ == $PROGRAM_NAME

File: day1/part2/solution.go
package main

import (
	"bufio"
	"fmt"
	"math"
	"os"
	"strconv"
)

// A helper for true mathematical floor division
func floorDiv(a, b int) int {
	return int(math.Floor(float64(a) / float64(b)))
}

func solve(rotations []string) int {
	currentPosition := 50
	totalZeroCount := 0

	for _, line := range rotations {
		if len(line) < 2 {
			continue
		}

		direction := line[0]
		distance, err := strconv.Atoi(line[1:])
		if err != nil {
			continue
		}

		zerosThisTurn := 0
		if direction == 'R' {
			zerosThisTurn = floorDiv(currentPosition+distance, 100) - floorDiv(currentPosition, 100)
		} else if direction == 'L' {
			zerosThisTurn = floorDiv(currentPosition-1, 100) - floorDiv(currentPosition-distance-1, 100)
		}
		totalZeroCount += zerosThisTurn

		if direction == 'R' {
			currentPosition += distance
		} else if direction == 'L' {
			currentPosition -= distance
		}
	}
	return totalZeroCount
}

func main() {
	inputFilePath := "day1/part1/input1.txt"
	file, err := os.Open(inputFilePath)
	if err != nil {
		fmt.Fprintf(os.Stderr, "Error opening file '%s': %v\n", inputFilePath, err)
		os.Exit(1)
	}
	defer file.Close()

	var rotations []string
	scanner := bufio.NewScanner(file)
	for scanner.Scan() {
		rotations = append(rotations, scanner.Text())
	}

	if err := scanner.Err(); err != nil {
		fmt.Fprintf(os.Stderr, "Error reading file: %v\n", err)
		os.Exit(1)
	}

	password := solve(rotations)
	fmt.Printf("The password is: %d\n", password)
}
